runtime/parachains/src/ump.rs

```diff
@@ -45,45 +44,56 @@ pub trait UmpSink {
 	/// it did not begin processing a message since it would otherwise exceed `max_weight`.
 	///
 	/// See the trait docs for more details.
-	fn process_upward_message(origin: ParaId, msg: &[u8], max_weight: Weight) -> Option<Weight>;
+	fn process_upward_message(origin: ParaId, msg: &[u8], max_weight: Weight) -> Result<Weight, (MessageId, Weight)>;
 }
 
 /// An implementation of a sink that just swallows the message without consuming any weight. Returns
 /// `Some(0)` indicating that no messages existed for it to process.
 impl UmpSink for () {
-	fn process_upward_message(_: ParaId, _: &[u8], _: Weight) -> Option<Weight> {
-		Some(0)
+	fn process_upward_message(_: ParaId, _: &[u8], _: Weight) -> Result<Weight, (MessageId, Weight)> {
+		Ok(0)
 	}
 }
 
+/// Simple type used to identify messages for the purpose of reporting events. Secure if and only
+/// if the message content is unique.
+pub type MessageId = [u8; 32];
```

Comment:
If we care about this invariant, should it be a unit type with its own constructor?